### PR TITLE
Microcontrollers management (Work in Progress)

### PR DESCRIPTION
**Pull request by astralien3000**
Hello !

Here is my PR for microcontroller management.
This not completed, but you can give me feedback.
I am using this work with my library [Aversive++ (branch 'dev-conan')](https://github.com/AversivePlusPlus/AversivePlusPlus/tree/dev-conan), if you want an example.

### CODE REVIEW

**Comment by memsharded** [comment]
Hi!
Some tests have failed in the CI. Don't worry I'll have a look and fix the tests if appropriate or give feedback otherwise. Also review all the related threads discussions, I have been disconnected for a while, sorry for that. Cheers!

**Comment by astralien3000** [comment]
Hmm, I did not check the tests before pushing, sorry !
Yes, no problem about your "disconnection" ^^, I am just a little impatient because it would help me to solve a problem that I face for several months. The problem is not so big, but it would give a cleaner project structure.

**Comment by astralien3000** [comment]
Hey ! I haven't fixed previous CI fails, but I added the `Embedded` class, as discussed in #86. What do you think of that solution ?

Here are some examples of use : 
 - https://github.com/AversivePlusPlus/hal-switch/blob/dev-conan/conanfile.py
 - https://github.com/AversivePlusPlus/hdl-atmegaxx0_1/blob/dev-conan/conanfile.py

**Comment by astralien3000** [comment]
Hey ! I've uploaded a proof of concept that makes external the embedded tree representation.
[Here is the repos](https://github.com/astralien3000/conan-embedded-settings) that contains the data.

The concept is simple, I use github's feature to download the zip archive of the repository, and unzip it in the `~/.conan` directory. Then the class EmbeddedSettingsParser can load yml files and behave as the previous version.

@memsharded @lasote, I have a code tagged "VERY DIRTY", I could not find where to put it in a clean way. Do you have any suggestion ?

TODO:
 - [ ] Move dirty code
 - [ ] Change files names in `~/.conan`
 - [ ] Make a proper external repository, and move it to conan-io organization

IMPROVEMENTS:
 - [ ] Be able to load several yml files
 - [ ] Generate settings (the `settings.target` part) from embedded settings

File: conans/client/embedded.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+from conans.paths import conan_expand_user
+from conans.client.paths import ConanPaths
+from conans.client.output import ConanOutput
+from conans import tools
+
+import yaml
+import os
+import sys
+
+class EmbeddedSettingsParser:
```
Comment (memsharded) [inline conans/client/embedded.py]: remember to use modern python classes: ``class EmbeddedSettingsParser(object):``

**Comment by memsharded** [comment]
Finally found some time to resume this. Quick question, from your Aversive example:

```
arm:

cortex_m4:
    parents: [arm]

stm32:
    parents: [arm]
stm32f4:
    parents: [stm32, cortex_m4]
```

What is the meaning of such diamond? Does it make sense that stm32f4 inherits from both stm32 and cortex_m4?

**Comment by astralien3000** [comment]
Well, I did not though of this, but to explain everything : 

 - Cortex-M4 is an ARM (armv7) processor, it can be embedded in microcontrollers chips.
 - STM32 is a family of microcontrollers, which are all ARM-based (contain an ARM processor, as far as I know, from the Cortex-M* family).
 - STM32F4 is a subfamily of STM32 that contains a Cortex-M4 processor.

The link between `cortex_m4` and `arm` is not the same as the one between `stm32` and `arm`, as you can see. And since any subfamily of STM32 have to inherit from a "Cortex-M*" processor, it is possible to remove `arm` from `stm32`'s parents, actually.

**Comment by memsharded** [comment]
Ok, I have spent quite a few hours looking at this, investigating and considering design and implementation issues. Sorry for such long delays, we are terribly busy. Your proposal has been really enlightening, and totally goes in the right direction. Very, very useful, lets summarize my current vision of the solution, mainly from the user point of view:

1. Only a new ``target: ANY`` setting would be required. Without values, ranges.... lets keep it simple. The new ``ANY`` value allows it to take any value without problems.

2. The hard processing will be done by an ``Embedded`` object helper, that takes a text definition of the known "ontology" of embedded targets: architectures, families, series, micros, boards... defined in a hierarchical convenient way. Such ``Embedded`` object also receives the ``settings`` as input, and can return properties as ``embed.arch``, ``embed.family``, ``embed.micro``, ``embed.board``. 

3. The ``target`` input setting can be anything, for example a ``family``. In such case, child fields like ``micro`` or ``board`` will be ``None``.

4. The ``target`` input can be easily narrowed or modified both pre and post build step. If we only need for building the architecture, it can be narrowed pre-build and the following could be done:
```python
def configure(self):
    embed = Embedded(self.settings)
    self.settings.target = embed.arch
```

Now, lets go to the implementation details. I think this has enough complexity to be developed with care, good testing, etc. So from my point of view we need:

1. Totally decouple the ``Embedded`` class from the rest of the application. It cannot be coupled to paths, output, etc. Its constructor will have 2 parameters: the text definition of the hierarchy, and the ``settings`` object.  It will dynamically create and provide the properties (``family``, ``arch``, etc) when parsing the definition. It will search from the ``target`` the corresponding match, which is not necessarily a ``board``, it can be anywhere in the hierarchy. 

2. Do good testing of the ``Embedded`` class, this is core to the application

3. Create a callable that will make available the ``Embedded`` object to the ``conanfiles`` that want to use it. This callable (factory?) will manage to get from the path the text definition of microprocessors from a file called for example ``targets.yml``. But it is important that this functionality is decoupled from the business logic. 

4. Define the origin, collaboration, sharing, updates of the ``targets.yml`` file. This is an important part, as evolving it can break existing packages, and not updating it when new packages use updated definitions will also produces failures. In any case, this is a decoupled problem, and would happen with any model, so we can think about it later.

What do you think? Do you think the model would be good enough? The implementation would require a bit of python trickery, especially if we don't want to hardcode properties and let them be defined by the data in ``targets.yml``, and setting up the tests. Do you want help with it? I can make some scaffolding.

@lasote, please some feedback would be welcome

Thanks again @astralien3000 for your great job.

**Comment by lasote** [comment]
I agree with @memsharded proposal and implementation details. I think the ANY setting is the right way and the new yml parsing idea from @astralien3000 and the Embedded class too. It solved the two main problems that I was not able to solve, the fixed settings (with the ANY) and the dynamic Embedded class. :+1: 
Thanks both for your great work! I can help with the implementation if you want.

**Comment by astralien3000** [comment]
Thank you ! ^^

> Do you want help with it? I can make some scaffolding.

Yes ! I begin to understand conan's internal structure, and would like to continue helping.

But I have some questions/remarks about your propositions : 

> Only a new target: ANY setting would be required. Without values, ranges.... lets keep it simple. The new ANY value allows it to take any value without problems.

I am not really against (since it would be the simplest thing to do). But I may propose to not add anything to `settings.yml`(not even the fact that there is a setting called `target`), but add the targets to settings "at runtime" from `target.yml`. So that the user cannot use invalid names, and conan can display proper errors at the right time. What do you think ?

> The target input setting can be anything, for example a family. In such case, child fields like micro or board will be None.

Hmm, for example, a target `stm32` would not make any sense, I think. All STM32 are not even the same ARM version (v6/v7). This is why I differentiate 2 types of targets to the others (which are more "groups" than real targets) : "microcontrollers" and "boards", that define everything needed to compile. To that you can add the "processor" or "arch" type, which is the same as what you call `settings.arch`.

> The target input can be easily narrowed or modified both pre and post build step. If we only need for building the architecture, it can be narrowed pre-build and the following could be done:
```yml
def configure(self):
    embed = Embedded(self.settings)
    self.settings.target = embed.arch
```

I am not sure of the use of this feature. Does it enable the libraries to choose itself the "final" target ? Quite dangerous, and error prone, in my opinion.

> Totally decouple the Embedded class from the rest of the application. It cannot be coupled to paths, output, etc. Its constructor will have 2 parameters: the text definition of the hierarchy, and the settings object. It will dynamically create and provide the properties (family, arch, etc) when parsing the definition. It will search from the target the corresponding match, which is not necessarily a board, it can be anywhere in the hierarchy. 

👍  Ok, `EmbeddedSettingsParser` will only parse text, which will be provided by something else. But adding some methods to, for example `ConanPaths`, like `embedded_settings_path` is ok, isn't it ?

> Create a callable that will make available the Embedded object to the conanfiles that want to use it. This callable (factory?) will manage to get from the path the text definition of microprocessors from a file called for example targets.yml. But it is important that this functionality is decoupled from the business logic. 

👍  [Like this ?](https://github.com/AversivePlusPlus/hal-switch/blob/a0d56b3feab0f044bc69fe479ebe61e04f232220/conanfile.py)

**Comment by memsharded** [comment]
I think there are two approaches:

1. Dynamic creation of settings from ``targets.yml``, which can create different settings, like ``target``, ``board``, ``micro`` and any other setting we might need. If the targets are already fully translated to settings, then, an ``Embedded`` class will not be necessary (maybe internally for the translation), just explicit use of settings.

2. Just a single static setting ``target`` and dynamic interpretation in the conanfiles through an ``Embedded`` object available for the package creators to extract information about families, architectures, etc.

I think in your approach you might be missing an important point, that is the narrowing of settings to common values. Let's say that you are building a mathematical logic library, ``PackageMatrix``, and all you need to know is the architecture of the target. But not the specific micro-controller, board, etc. If you don't narrow the input settings to the architecture, then you will have a ``PackageMatrix`` different binary for each specific micro-controller, which is totally unnecessary.

So, at some point it is necessary to specify in the package which settings must be taken into account. This can be done:

1. With the dynamic creation of settings, it will be declaring the settings:

  ```python
  class PackageMatrix(Conanfile):
      settings = "target_architecture", "target_family"  # target_micro is not necessary for this package
  ```

2. With the static ``target`` setting, we need the help of the ``Embedded`` object to compute it dynamically:

  ```python
  class PackageMatrix(Conanfile):
      def configure(self):
          self.settings.target = Embedded(self.settings).target_architecture
  ```

In any case, it is the package creator who decides which settings affect their package binary output.

The problem with the "groups" is that there are some groups that really make sense to define a binary, and otherw doesn't make sense. It would be great if we could restrict the ``targets`` definition to just those settings required to build the binary. If ``stm32`` doesn't help at all to decide how to build a binary, then maybe it shouldn't be in the model.


Yes, the idea for the callable is to be used exactly as you use it, but in order to provide it with the paths and output from conan, an object has to be created first, which can be called with ``self.settings`` and return in turn the real ``Embedded`` object. It is a common pattern for dependency injection, don't worry, I can help with that setup.

There are pros & cons for both 1 and 2 strategies, that was what I was analyzing the other day:

1. Pros:
   - Typical conan settings approach, they are dynamically defined, but for the users it is mostly the same. 
  - No need to learn the semantics of a new ``Embedded`` object, it will be already provided by the settings.

   Cons:
  - Might be a bit trickier to implement, managing complex settings it is not simple


2. Pros:
   - Very simple & static settings setup, just ``target: ANY``
   - Powerful semantics can be added to ``Embedded`` logic, that can help for package creators.
   - Maybe easier to implement

   Cons:
   - Might be more error-prone on the user (package-creator) side.

I was slightly more balanced towards 2, but to be fair, just slightly, maybe 55%-45%. If you think that the approach to narrowing will be too much error prone for 2, we might re-consider approach 1. Thanks very much once more! :)

**Comment by astralien3000** [comment]
Ah, ok, I understand better !
To be completely clear, with your proposition (2), let's consider this example : 

 - A library called `A` is the same for all microcontrollers of the STM32F4 family. It is hardware independent (that is to say that it does not use MCU specific registers and functionality). So that the binary archive will be the same for all MCUs (every STM32F4 is a Cortex-M4)
 - An other library called `B` supports also all the STM32F4 family, but is hardware dependent. So that the library needs to be compiled for each MCU.
 - The user want to compile a project for a STM32-Discovery board (which use a STM32F407VG micro) that use libraries `A` and `B`.

The conanfile.py of `A` would look like : 
```python
class MyConanFile(ConanFile):
    settings = "target"

    def configure(self):
        embed = Embedded(self.settings)
        self.settings.target = embed.family
```

The conanfile.py of `B` would look like : 
```python
class MyConanFile(ConanFile):
    settings = "target"

    def configure(self):
        embed = Embedded(self.settings)
        self.settings.target = embed.micro
```

And the user would type a command like that : 
```bash
conan install -s target=stm32f4_discovery .
```

So that, for package `A`, `settings.target = cortex_m4`, and for `B`, `settings.target = stm32f407vg`, but for the user project, `settings.target = stm32f4_discovery`.

It is a little weird for what I understood from settings, but if it enables to compile the library only once for a whole arch/family/group, I think it is cool ! Is it already possible to do that currently ? I mean, to change the settings of a package ? For example, if a package is the same for Linux and MacOS, but different for windows, you can just do : 
```python
    def configure(self):
        if self.settings.os == "Macos":
            self.settings.os = "Linux"
```

If I understood everything well, I quite agree with (2). Let's consider that the package creator knows what he is doing !

**Comment by lasote** [comment]
@memsharded The method to use in the `conanfile` should not be `conan_info` instead of `configure`?

**Comment by memsharded** [comment]
@lasote If you don't need the full settings, you can narrow them pre-build in the ``configure()`` step too.

@astralien3000 Yes, you understood it right! That is the way current settings work, and they allow exactly that, to have one library built for the "minimum common denominator". For example a header-only library, that does not produce binaries, can fully remove the settings. It will only have one package for any configuration, containing just the headers.

About the OS change, not exactly, you cannot change the OSs you are in. Even if the package code (conanfile.py) is the same for Linux and MacOSX, the resulting binary will be different! So the binary for Linux is not the binary for OSX. Just develop your package code taking this into account, but dont tell the system the package is for Linux if you are in OSX, because it will break in Linux (it will retrieve a OSX built binary)

```python
def build(self):
    if self.settings.os != "Windows":
        # Just do some windows build stuff
   else:
       # Do *nix build stuff
```

Note that for the settings I was suggesting for micros, you are not changing the system, you are narrowing the settings, but they have to be true and really match the platform.

**Comment by astralien3000** [comment]
Hello,
Well, my project using conan for microcontrollers is kind of dead. So that I'm likely to stop using conan for a long time (if not definitely... I don't know what the future is made of). I can still try to help you take the plunge. Or I should close this PR.

**Comment by memsharded** [comment]
Hi @astralien3000,

very sad to hear that your project is kind of dead. The truth is that this PR has been standing for a long time, we haven't closed it (and we have done it for other PRs), because we strongly believe that there is good work in the good direction in it. We apologize for not being able to move this forward the way it deserves, conan growth (mostly for OS based, not embedded) has kept us 200% busy, and not being active in the embedded field ourselves makes it more difficult. 

Please, do not close it, there is valuable work that we don't want to lose. We don't know when, but we are sure that we will end using it, if you are still able to help, that would be great.

Thanks again once more for your contribution and help.

**Comment by astralien3000** [comment]
Thanks ^^

Ok, no problem, I will keep it open. Anyway, closing it would not remove my work. It would just be harder to find. You also may want to copy my branch, and clone my [external repository](https://github.com/astralien3000/conan-embedded-settings), to keep a copy just in case, even if I don't plan to delete it.

Also ! I did have a thought about this PR some time ago. I was thinking that `target` and `arch` settings may be exactly the same thing. So that I would suggest to use arch instead of target. But I haven't tried it.

Your project is really cool ! I hope I will have a new occasion to use it in he future ! Keep going ! ^^

**Comment by shreyasbharath** [comment]
A quick question since it is somewhat relevant to this PR.

How do I use Conan to package a pre-compiled lib from a custom cross-compiler (targeting a Renesas MCU for example)? Does it have to be defined inside Conan or can I define it myself?

**Comment by memsharded** [comment]
@shreyasbharath

You might want to read how to package existing binaries: http://docs.conan.io/en/latest/packaging/existing_binaries.html#packaging-pre-built-binaries

And then, you might want to customize the standard settings to add your own platforms and architectures: http://docs.conan.io/en/latest/faq/troubleshooting.html#error-invalid-setting

**Pull request closed without merging**